crates/bevy_app/src/plugin_group.rs

```diff
@@ -11,25 +11,55 @@ struct PluginEntry {
     enabled: bool,
 }
 
+/// Builds and customizes a plugin group. A plugin group is an ordered list of plugins that
+/// that can be enabled, disabled or reordered.
 #[derive(Default)]
 pub struct PluginGroupBuilder {
     plugins: HashMap<TypeId, PluginEntry>,
     order: Vec<TypeId>,
 }
 
 impl PluginGroupBuilder {
+    // Removes a previous ordering of a plugin that has just been added at `added_at` index
+    fn remove_when_adding<T: Plugin>(&mut self, added_at: usize) {
+        if let Some(to_remove) = self
+            .order
+            .iter()
+            .enumerate()
+            .find(|(i, ty)| *i != added_at && **ty == TypeId::of::<T>())
+            .map(|(i, _)| i)
+        {
+            self.order.remove(to_remove);
+        }
+    }
+
+    /// Adds the plugin `plugin` at the end of this `PluginGroupBuilder`. If the plugin was
+    /// already in the group, it is removed from its previous place.
     pub fn add<T: Plugin>(&mut self, plugin: T) -> &mut Self {
+        let target_index = self.order.len();
         self.order.push(TypeId::of::<T>());
-        self.plugins.insert(
+        if let Some(entry) = self.plugins.insert(
```

Comment:
This new logic is now duplicated in three places inside `PluginGroupBuilder`. Would it be possible to for example move it into a private function?